ide/projectapi/src/org/netbeans/api/project/ProjectActionContext.java

```diff
@@ -0,0 +1,402 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import org.netbeans.api.annotations.common.CheckForNull;
+import org.netbeans.api.annotations.common.NonNull;
+import org.netbeans.api.project.Project;
+import org.netbeans.spi.project.ProjectConfiguration;
+import org.openide.util.Lookup;
+import org.openide.util.lookup.Lookups;
+import org.openide.util.lookup.ProxyLookup;
+
+/**
+ * This represents context for a project model query. The context describe a project action
+ * for which the query should be evaluated, project configuration (if not the active one), 
+ * and possible custom overrides.
+ * {@link ProjectConfiguration} or the IDE-supported project action and the associated properties, profiles, or Lookup may
+ * affect behaviour of queries across the IDE.
+ * <p>
+ * Properties map to user-settable properties of the build system, such as system properties in Maven
+ * or project properties in Gradle.
+ * <p>
+ * Profiles map to profiles in Maven, but do not have any effect in Gradle at the moment; this might change.
+ * <p>
+ * Instances of {@link ProjectActionContext} may be passed to some queries as (part of the) parameters explicitly. In case
+ * the API does not support such flexibility directly, the implementation(s) may - the context may be published for the
+ * query computation using {@link #apply}; the implementation may then use {@link #find} to obtain a {@link ProjectActionContext}
+ * effective for that project.
+ * <p>
+ * <b>Important note:</b> Not all project queries support {@link ProjectActionContext}. Queries which do should mention that
+ * fact in their documentation, or expose it in their APIs.
+ * 
+ * @author sdedic
+ * @since 1.89
+ */
+public final class ProjectActionContext {
+    private final Project project;
+
+    ProjectActionContext(Project project) {
+        this.project = project;
+    }
+   
+    /**
+     * The project action.
+     */
+    private String projectAction;
+    
+    /**
+     * The desired project configuration
+     */
+    private ProjectConfiguration configuration;
+    private Map<String, String> properties;
+    private Set<String> profiles;
+    private Lookup lookup;
+
+    /**
+     * Returns the project this context applies to. Having {@link Project} as a property allows multiple
+     * contexts, individually for each project, to be active.
+     * @return the target project.
+     */
+    public @NonNull Project getProject() {
+        return project;
+    }
+
+    /**
+     * Additional information for the underlying project implementation, similar to 
+     * context lookup in {@link org.netbeans.spi.project.ActionProvider#invokeAction}.
+     * Null if no specific info is present.
+     * 
+     * @return Lookup with additional information or {@code null}.
+     */
+    public @CheckForNull Lookup getLookup() {
+        return lookup;
```

Comment:
Can return Lookup.EMPTY

Reply:
Hm, with all those nulls (including this one), I intended to signal explicitly "no value". For example, if no Lookup + properties + profiles are present the implementation can take a great shortcut (= no customization to the project model or services).

Reply:
Well if those shortcuts are important, then probably I'd intorduce a ```"boolean isShortcutAllowed()"``` method here to do the necessary checks inside here. Of course I assume that you can do shortcuts if no lookup + properties + profiles are present.

Reply:
The little trouble is that it is the project implementation that decides which part may potentially affect the project data evaluation ... all stuff except the Lookup can be checked for 'no ops' (empty), the Lookup is ... well, peculiar. Eventually I could document that the exact value `Lookup.EMPTY` may be used for optimizations.

Reply:
Addressed in a1c2f1b8c6b2